executor/linux_fork_server_executor.cpp

```diff
@@ -0,0 +1,585 @@
+    CreateJoinedEnvironmentVariables( std::move( environment_variables_ ) );
+
+    put_channel.SetupForkServer((char* const*) cargv.data());
+}
+
+/**
+ * Postcondition:
+ *  - Free resources handled by this class, then invalidate data.
+ *      - Close input_fd file descriptor. then the value is invalidated (fail-safe)
+ *      - If running in fork server mode, close the pipes for communicating with fork server, then terminate fork server process.
+ *  - Temporary purpose: If actve_instance which signal handlers refer has a pointer to self value, assign nullptr to it and invalidate.
+ */
+LinuxForkServerExecutor::~LinuxForkServerExecutor() {
+    // Since active_instance can refer another instance of multiple N, it is considered FuzzerHandle::Reset is called if active_instance is not self value, therefore do nothing. On the other hand, assign nullptr if active_instance is self value.
+    // Although, assigning nullptr just in case, since the handlers will never called when the destructor is called, it is basically not a problem.
+    if (active_instance == this) {
+        active_instance = nullptr;
+    }
+
+    if (input_fd != -1) {
+        Util::CloseFile(input_fd);
+        input_fd = -1;
+    }
+
+    if (null_fd != -1) {
+        Util::CloseFile(null_fd);
+        null_fd = -1;
+    }
+
+    EraseSharedMemories();
+
+    TerminateForkServer();
+    // Although, PUT process should be handled by fork server, kill it just in case. (Since fork server is expected to be killed using kill, therefore it will never becoome a zombie. So never wait.)
+    KillChildWithoutWait();
+}
+
+void LinuxForkServerExecutor::SetCArgvAndDecideInputMode() {
+    assert(!argv.empty()); // It's incorrect too far
+
+    stdin_mode = true; // if we find @@, then assign false to stdin_mode
+
+    for (const auto& v : argv ) {
+        if ( v == "@@" ) {
+            stdin_mode = false;
+            cargv.emplace_back(path_str_to_write_input.c_str());
+        } else {
+            cargv.emplace_back(v.c_str());
+        }
+    }
+    cargv.emplace_back(nullptr);
+}
+
+void LinuxForkServerExecutor::TerminateForkServer() {
+    put_channel.TerminateForkServer();
+}
+
+/**
+ * An static method
+ * Precondition:
+ *  - active_instance has a non-nullptr value definitely when this function is called.
+ *  - active_instance has an address of valid Executor instance.
+ *  - The global timer exists in the self process, and the function is called for SIGALRM signal.
+ * Postcondition:
+ *  - The function works as signal handler for SIGALRM
+ *      - If called, the PUT is considered as expiring the execution time, therefore kill active_instance->child_pid.
+ *      - Set a flag active_instance->child_timed_out
+ */
+void LinuxForkServerExecutor::AlarmHandler(int signum) {
+    assert (signum == SIGALRM);
+    assert (active_instance != nullptr);
+    active_instance->KillChildWithoutWait();
+    active_instance->child_timed_out = true;
+}
```

Comment:
I think this is no longer needed

Reply:
Thank you for suggestion. I removed these lines.